crates/crypto/src/threshold_signature/indexed/signature.rs

```diff
@@ -0,0 +1,219 @@
+/// an `IndexedSignature`. Verification uses that index to fetch the expected public key and
+/// compare it against the recovered key from the signature.
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct IndexedSignature {
+    /// Index of the signer in the ThresholdConfig keys array (0-255).
+    index: u8,
+    /// 65-byte recoverable ECDSA signature (header || r || s).
+    ///
+    /// The header byte contains the recovery ID, possibly with BIP-137 address type encoding.
+    /// See struct-level documentation for format details.
+    signature: [u8; 65],
+}
+
+impl IndexedSignature {
+    /// Create a new indexed signature.
+    pub fn new(index: u8, signature: [u8; 65]) -> Self {
+        Self { index, signature }
+    }
+
+    /// Get the signer index.
+    pub fn index(&self) -> u8 {
+        self.index
+    }
+
+    /// Get the header byte (first byte of the signature).
+    ///
+    /// This byte contains the recovery ID, possibly encoded in BIP-137 format.
+    /// The verification code handles normalization automatically.
+    ///
+    /// # Format
+    /// - Raw: 0-3 (recovery ID directly)
+    /// - BIP-137: 27-42 (encodes address type + recovery ID)
+    pub fn recovery_id(&self) -> u8 {
+        self.signature[0]
+    }
+
+    /// Get the r component (bytes 1-32).
+    pub fn r(&self) -> &[u8; 32] {
+        self.signature[1..33]
+            .try_into()
+            .expect("signature[1..33] is always 32 bytes")
+    }
+
+    /// Get the s component (bytes 33-64).
+    pub fn s(&self) -> &[u8; 32] {
+        self.signature[33..65]
+            .try_into()
+            .expect("signature[33..65] is always 32 bytes")
+    }
+
+    /// Get the compact signature (r || s) without recovery ID.
+    pub fn compact(&self) -> [u8; 64] {
+        let mut compact = [0u8; 64];
+        compact.copy_from_slice(&self.signature[1..65]);
+        compact
+    }
+}
+
+/// A set of indexed ECDSA signatures for threshold verification.
+///
+/// Signatures are sorted by index and must not contain duplicates.
+#[derive(Debug, Clone, PartialEq, Eq, Default, BorshSerialize, BorshDeserialize)]
+pub struct SignatureSet {
+    /// Sorted signatures by index, no duplicates.
+    signatures: Vec<IndexedSignature>,
+}
+
+impl SignatureSet {
+    /// Create a new signature set from a vector of indexed signatures.
+    ///
+    /// The signatures will be sorted by index and checked for duplicates.
+    pub fn new(mut signatures: Vec<IndexedSignature>) -> Result<Self, ThresholdSignatureError> {
+        // Sort by index
+        signatures.sort_by_key(|s| s.index);
+
+        // Check for duplicate indices
+        for window in signatures.windows(2) {
+            if window[0].index == window[1].index {
+                return Err(ThresholdSignatureError::DuplicateSignerIndex(
+                    window[0].index,
+                ));
+            }
+        }
```

Comment:
Recommend using an internal hash set to check for duplicates instead. It's a very clean approach.